build.wake

```diff
@@ -15,57 +15,87 @@ def macrosParadiseIvyDep =
   readIvyDepsJSON here
   | rmapPass extractIvyDep
 
-# TODO Should this be in the standard lib?
-def orElse b = match _
-  Some a = Some a
-  None   = b
-
 # Adds an a compiler plugin dependency on macros paradise
 # If module is already broken, keeps old error
 global def addMacrosParadiseCompilerPlugin module =
   match macrosParadiseIvyDep
     Pass dep = module | editScalaModuleCompilerPlugins (dep, _)
     Fail err = module | editScalaModuleError (orElse (Some err))
 
+# TODO This should be default behavior in api-scala-sifive
+def scalaModuleAddExtraSourceDirs module =
+  def major = module.getScalaModuleScalaVersion.scalaVersionMajorString
+  def dir = "src/main/scala-{major}"
+  module
+  | editScalaModuleSourceDirs (dir, _)
+
+# NOTE, currently unused, see addChisel3CompilerPlugin
+global def chisel3Plugin =
+  makeScalaModuleFromJSON here "chisel3Plugin"
+  | setScalaModuleRootDir "{chisel3Root}/plugin"
+  | scalaModuleAddExtraSourceDirs
+
+# We ignore scalaVersion because it comes from ScalaModule
+# The scalaVersion in ivydependencies.json is purely for fetching
+def chisel3PluginIvyDep =
+  def extractIvyDep json =
+    json // `chisel3PluginIvyDep` // `dependencies` | getJString
+    | omapPartial stringToIvyDep
+    | getOrFail (makeError "chisel3Plugin ivy dependency not found!")
+  readIvyDepsJSON here
+  | rmapPass extractIvyDep
```

Comment:
```suggestion
def chisel3PluginIvyDep =
  require Pass ivyDepsJSON = readIvyDepsJSON here
  require Some depString = ivyDepsJSON // `chisel3PluginIvyDep` // `dependencies` | getJString
    else Fail (makeError "chisel3Plugin ivy dependency not found in ivydependencies.json!")
  require Some dep = stringToIvyDep depString
    else Fail (makeError "Invalid chisel3PluginIvyDep dependency string: {depString}")
  Pass dep
```

Here's how you could write it in a more imperative form with `require`.

Also, nit: the error message you wrote is misleading for the case where there's a typo in the dependency string format, since it's not that the dependency is missing, but that the dependency string was invalid. Also, I modified your other error message to include the `ivydependencies.json` file, since that will help point people who are not familiar with this code to the source of the error.

Reply:
I DRY-ed out this code (the `chisel3PluginIvyDep` code was copied), and converted it to your suggestion, thanks!